### PR TITLE
L2 -> L1 messages

### PR DESCRIPTION
**Pull request by war-in**
image

### CODE REVIEW

File: src/trace_reader.rs [hunk 1]
```diff
@@ -153,6 +162,12 @@ impl ResourcesKeys {
                 unit: context.string_id(&unit_string).into(),
             });
         }
+        for data in &self.onchain_data {
+            value_types.push(ValueType {
+                r#type: context.string_id(data).into(),
+                unit: context.string_id(&" payload length".to_string()).into(),
+            });
+        }
```
Comment (piotmag769) [inline src/trace_reader.rs:170]: This `for` won't work for storage costs (unit)

**Review by piotmag769 — APPROVED** [review]

File: src/trace_data.rs [hunk 2]
```diff
@@ -174,3 +175,8 @@ impl ExecutionResources {
         true
     }
 }
+
+#[derive(Clone, Default, Debug, Serialize, Deserialize)]
+pub struct OnchainData {
+    pub l2_l1_message_sizes: Vec<usize>,
+}
```
Comment (piotmag769) [inline src/trace_data.rs:182]: Remember to reuse this one in forge after https://github.com/foundry-rs/starknet-foundry/pull/1766 is merged

**Review by piotmag769 — APPROVED** [review]

**Comment by MaksymilianDemitraszek** [comment]
I believe it should be `sent payload length`

File: src/trace_reader.rs [hunk 3]
```diff
@@ -31,6 +31,7 @@ pub struct Sample {
     pub location: Location,
     pub sample_type: SampleType,
     pub flat_resources: ExecutionResources,
+    pub onchain_data: OnchainData,
```
Comment (MaksymilianDemitraszek) [inline src/trace_reader.rs:34]: Not sure if this is a good name, this field does not store `onchain_data`. `l1_resources` maybe?

File: src/trace_reader.rs [hunk 4]
```diff
@@ -131,6 +139,7 @@ pub fn collect_samples_from_trace(trace: &CallTrace) -> Vec<Sample> {
 pub struct ResourcesKeys {
```
Comment (MaksymilianDemitraszek) [inline src/trace_reader.rs:139]: I know it has not been introduced by you, but what this structure stores? I don't understand from the name.

Comment (piotmag769) [reply src/trace_reader.rs:139]: It introduces the keys used to create the resource list needed by pprof

Comment (MaksymilianDemitraszek) [reply src/trace_reader.rs:139]: what are 
> keys used to create the resource list needed by pprof

Comment (piotmag769) [reply src/trace_reader.rs:139]: https://github.com/google/pprof/tree/main/proto#measurement-values

Comment (MaksymilianDemitraszek) [reply src/trace_reader.rs:139]: Yes, but why do we have a structure with three fields without a single hash set? Aren't those fields isomorphic?

Comment (piotmag769) [reply src/trace_reader.rs:139]: https://github.com/software-mansion/cairo-profiler/pull/38

**Pull request merged**